### PR TITLE
[MRG+1] Add warning context

### PR DESCRIPTION
**Pull request by alanbernstein**
Thanks for the feedback. Again forgive my ignorance with the workflow here: I reverted the first commit, made a feature branch from that, and created a PR to matplotlib/v2.x.

Your suggestion makes more sense. My first attempt was closer to that, but I was trying to preserve the original warning attributes, and went down the wrong path.

https://github.com/matplotlib/matplotlib/issues/7601

### CODE REVIEW

File: lib/matplotlib/style/core.py [hunk 1]
```diff
@@ -188,7 +188,13 @@ def read_style_directory(style_dir):
     """Return dictionary of styles defined in `style_dir`."""
     styles = dict()
     for path, name in iter_style_files(style_dir):
-        styles[name] = rc_params_from_file(path, use_default_template=False)
+        with warnings.catch_warnings(record=True) as warns:
+            styles[name] = rc_params_from_file(path, use_default_template=False)
```
Comment (tacaswell) [inline lib/matplotlib/style/core.py]: This line is too long, we enforce pep8 via tooling in the tests.

**Comment by tacaswell** [comment]
Can you interactively rebase this to squash out the first two commits and then force-push the single commit up to your 2.x branch? `git rebase -i`

Ping me if you need help with that.

**Comment by QuLogic** [comment]
After doing the interactive rebase; do `git push --force origin add-warning-context`. If you don't force push and follow git's advice to `git pull`, you will end up with two copies of the commits; one from before you rebased and one from after.

You don't need to do a full rebase to fix this; just reset back to the correct new commits:
```bash
$ git branch
# Make sure you are on the 'add-warning-context' branch.
$ git status
# Check you haven't got any leftover changes you want to save with git stash or git commit on another branch.
$ git reset --hard b12ea701712fdb6d2ff9b8943aa6fd9a28b89b38
$ git push --force origin add-warning-context
```

**Comment by tacaswell** [comment]
:+1: 

Could you add test `matplotlib/tests/test_styles.py` ?  

I suggest using `tempdir` to make a temporary directory, writing out a know-to-be-bad style file and then using the same warning context manager to check that the warning contains the file name.

Adding this test means that in the future someone won't accidentally break the improvement you have added.

**Comment by NelleV** [comment]
This patch doesn't work on python3
@alanbernstein do you think you could fix the issue?

Else, it looks good.

**Comment by tacaswell** [comment]
I re-targeted this PR at the master branch and will merge it once it passes will merge it and leave it tagged as 2.0.1.

This should be backported after we have 2.0 final out.

**Comment by QuLogic** [comment]
Backported to `v2.0.x` as 4846dea41d9f6e491112a3ed43dca6e19bed6865.

**Pull request merged**